distributed/tests/test_client.py

```diff
@@ -1550,6 +1551,38 @@ def g():
                 break
 
 
+@gen_cluster(client=True)
+def test_upload_file_egg(c, s, a, b):
+    def g():
+        import package_1, package_2
+        return package_1.a, package_2.b
+
+    for value in [123, 456]:
+        dir_contents = {
+                'setup.py': (
+                    'from setuptools import setup, find_packages\n'
+                    'setup(name="my_package", packages=find_packages())\n'
+                ),
+                'package_1/__init__.py': 'a = {}'.format(value),
+                'package_2/__init__.py': 'b = {}'.format(value),
+        }
+        with tmp_dir(dir_contents) as directory:
+            subprocess.check_call(['python', 'setup.py', 'bdist_egg'], cwd=directory)
+            egg_root = os.path.join(directory, 'dist')
+            print(os.listdir(directory))
+            print(os.listdir(os.path.dirname(egg_root)))
+            print(os.listdir(egg_root))
+            egg_name = [fname for fname in os.listdir(egg_root) if fname.endswith('.egg')][0]
+            egg_path = os.path.join(egg_root, egg_name)
+            yield c.upload_file(egg_path)
+
+            x = c.submit(g, pure=False)
+            result = yield x
+            assert result == (value, value)
+    # this test won't impact the others, because the dir containg the
+    # egg is gone at this point.
```

Comment:
Just to be clear, will this affect the user's environment?  After running this test will a user find that they have  new packages named `package_1` and `package_2`?

Reply:
Each worker modifies their own `sys.path`. This change does not persist after the worker's process dies (they get closed by `gen_cluster`).

The only persisting side-effect is that [it writes a file to the worker's `local_dir`](https://github.com/dask/distributed/blob/master/distributed/worker.py#L688). I don't know if we care about this.

Reply:
Wait... Are the workers spawned by `gen_cluster` running in different processes?

Reply:
No, unless you provide the `Worker=Nanny` keyword argument.